--- nginx-1.0.11/src/event/ngx_event_connect.c	2011-09-23 03:32:16.000000000 +0400
+++ nginx-1.0.11/src/event/ngx_event_connect.c.new	2014-02-24 07:58:30.000000000 +0400
@@ -9,9 +9,8 @@
 #include <ngx_event.h>
 #include <ngx_event_connect.h>
 
-
 ngx_int_t
-ngx_event_connect_peer(ngx_peer_connection_t *pc)
+ngx_event_connect_peer(ngx_peer_connection_t *pc, struct timeval *timeout)
 {
     int                rc;
     ngx_int_t          event;
@@ -161,6 +160,33 @@ ngx_event_connect_peer(ngx_peer_connecti
 
             return NGX_DECLINED;
         }
+	else {
+	    if (timeout != NULL) {
+		fd_set sel;
+		long ret;
+		
+		FD_ZERO(&sel);
+		FD_SET((unsigned) s, &sel);
+		
+		ret = select(s + 1, NULL, &sel, NULL, timeout);
+		if (ret < 0) {
+		    ngx_log_error(NGX_LOG_CRIT, c->log, 0,
+			"set timeout via select() failed");
+		    //ngx_blocking(s);
+		    
+		    return NGX_DECLINED;
+		}
+		if (ret == 0) {
+		    ngx_log_error(NGX_LOG_INFO, c->log, 0,
+			"connect() to %V timed out", pc->name);
+		    //ngx_blocking(s);
+		    ngx_free_connection(c);
+		    ngx_close_socket(s);
+		    
+		    return NGX_ETIMEDOUT;
+		}
+	    }
+	}
     }
 
     if (ngx_add_conn) {
@@ -235,7 +261,7 @@ ngx_event_connect_peer(ngx_peer_connecti
     ngx_log_debug0(NGX_LOG_DEBUG_EVENT, pc->log, 0, "connected");
 
     wev->ready = 1;
-
+    
     return NGX_OK;
 
 failed:
