$dukzcry$
--- nginx-1.0.11/src/event/ngx_event_mutex.c	2011-09-23 03:32:16.000000000 +0400
+++ nginx-1.0.11/src/event/ngx_event_mutex.c.new	2014-02-24 07:58:30.000000000 +0400
@@ -9,11 +9,11 @@
 #include <ngx_event.h>
 
 
-ngx_int_t ngx_event_mutex_timedlock(ngx_event_mutex_t *m, ngx_msec_t timer,
-                                    ngx_event_t *ev)
+ngx_int_t ngx_event_mutex_lock(ngx_event_mutex_t *m, ngx_event_t *ev,
+				ngx_msec_t timer, ngx_log_t *log)
 {
-    ngx_log_debug2(NGX_LOG_DEBUG_EVENT, ev->log, 0,
-                   "lock event mutex %p lock:%XD", m, m->lock);
+    ngx_log_debug2(NGX_LOG_DEBUG_EVENT, log, 0,
+                   "lock event mutex %p lock: %XD", m, m->lock);
 
     if (m->lock) {
 
@@ -31,7 +31,8 @@ ngx_int_t ngx_event_mutex_timedlock(ngx_
         ev->light = 1;
 #endif
 
-        ngx_add_timer(ev, timer);
+	if (timer)
+    	    ngx_add_timer(ev, timer);
 
         return NGX_AGAIN;
     }
@@ -48,7 +49,7 @@ ngx_int_t ngx_event_mutex_unlock(ngx_eve
 
     if (m->lock == 0) {
         ngx_log_error(NGX_LOG_ALERT, log, 0,
-                      "tring to unlock the free event mutex %p", m);
+                      "trying to unlock the free event mutex %p", m);
         return NGX_ERROR;
     }
 
